### ISSUE TITLE
Discover the Hidden Truth: Apple's GPU MPS - A Disappointing Waste of Time and Resources!

### ISSUE THREAD

**Issue by traderpedroso** [guillaume-be/rust-bert#372]
Apple's M1 GPU MPS (Metal Performance Shaders) is not an ideal solution for training or making inferences with models that require float64 precision. When compared to a CPU, its performance is subpar, even when tailored to MPS standards. Consequently, the resulting predictions lack the accuracy associated with float64 precision.

So far, I've only managed to perform basic calculations. Has anyone had success with any models?

```rust
use tch::{kind, Tensor};

fn grad_example() {
    let mut x = Tensor::from(2.0f32)
        .to_device(tch::Device::Mps)
        .set_requires_grad(true);
    let y = &x * &x + &x + 36;
    println!("y {}", y.double_value(&[]));

    x.zero_grad();
    y.backward();

    let dy_over_dx = x.grad();
    println!("dy/dx {}", dy_over_dx.double_value(&[]))
}

fn main() {
    let t = Tensor::of_slice(&[3, 1, 4, 1, 5]);
    t.print(); // works on CPU tensors

    println!("t(cpu) {:?}", &t);
    println!("t device: {:?}", &t.device());
    let t = Tensor::randn(&[3, 1, 4, 1, 5], kind::FLOAT_CPU).to_device(tch::Device::Mps);
    t.print();
    println!("t(mps) {:?}", &t);
    println!("t device: {:?}", &t.device());

    grad_example();
}```

**Comment by guillaume-be** [comment]
Hello,

I don't think the example provided above can be used to provide a representative estimate of the benefits from using accelerators (MPS and CUDA devices included), the matrices are so small (and operations so simple) that a significant portion of the time probably is spent to the allocation stage and cross-device communication.

**Comment by traderpedroso** [comment]
I'm sorry, that was a basic example of using MPS. I'm sure I can achieve better performance training with MPS in different models, particularly with TensorFlow. However, the prediction performance was significantly lower compared to the same model trained on ROCm and CUDA, as well as when using the CPU. Of course, There are difficulties with data types due to limitations, as described in the document below.

For a more extensive list of which data types do and don't run:

Avoid Float64 on all Apple devices. Even if the hardware supports double physically (AMD or Intel), the Metal API doesn't let you access it.
Avoid BFloat16. It is natively supported by the latest Nvidia GPUs, but not supported in Metal. Also, don't try to use TF18/TF32 or Int4.
All standard integer types (UInt8, UInt16, UInt32, UInt64) and their signed counterparts work natively on Apple devices. Not exactly 8-bit integers, which are cast to 16-bit integers before being stored into registers, but those aren't going to harm performance. Yes, they run 64-bit integers on the Apple GPU and not 64-bit floats. Metal allows you to use 64-bit integers in shaders on AMD and Intel, but the arithmetic there might just happen through emulation (slow). I think that's where I experienced the crash in MPSGraph previously - trying to run an operation on UInt64 on my Intel iMac and on my hackintosh .
https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf


Of course, there are some tweaks that, with certain models, allow us to achieve better performance in terms of both speed and accuracy. For example, in the case of zero-shot classification using CUDA and rust-bert, I was able to achieve up to 7x better performance compared to the same situation using Python, and even up to 10x better performance when using ONNX Runtime with Rust, during the inference process. Employing techniques such as streaming and batching can further enhance these results.

I would like to make it clear that this is my point of view and does not define MPS. Even considering my technical limitations, I couldn't give a verdict on MPS. However, without a doubt, I believe at this stage it is a waste of time to expend efforts on MPS support.
